### DISCUSSION TITLE
On-device training of small LLM LoRA weights

### DISCUSSION THREAD

**Discussion by martinkorelic** [microsoft/onnxruntime#21447] [Training Q&A]
Bit of a beginner with the onnx runtime framework, but I noticed that it is one of the rare frameworks that allow on-device retraining of models. So here’s my question:

Does a framework like ONNX runtime offer such a thing for LLM models, by converting from existing transformer models and then finetuning them on device? Instead of re-training the whole model, we would only finetune the LoRA weights.

I understand the hardware restrictions of edge devices, but I would still like to know if such a thing is still feasible with this framework.

Also is it possible that on-device learning can be used with the underlying device GPU (like for Android…)?

Open to any answers, tips or discussion about this. Thank you!

**Comment by carzh** [comment]
On-device training does provide regular finetuning abilities by enabling users to freeze most layers and unfreeze some layers.  AFAIK, we haven't yet tested on-device training with finetuning the LoRA weights, but it should be possible with the following:

* Apply LoRA changes and config to PyTorch / transformers model
* Export the model using an exporter. 
* Generate the training artifacts. Mark all the lora linear weights as requires_grad, everything else as frozen
* Continue the rest of the training as per usual. 

If you give it a try, I'd be curious to hear how it goes. 

For your second question, on-device training is currently enabled on CUDA, but not other device GPUs.

**Reply by martinkorelic** [reply]
Is GPU training limited by the framework or is it possible to implement such a thing by using [ONNX runtime execution provider](https://onnxruntime.ai/docs/execution-providers/NNAPI-ExecutionProvider.html)? Or are they only used for inference at this point?

**Reply by carzh** [reply]
We're limited by op support for a given execution provider. Gradient ops would have to be implemented to train on the device GPU. Otherwise, it'll fall back to the CPU gradient op implementation.

**Comment by ColdCodeCool** [comment]
@martinkorelic  sorry to reply a closed issue, but I wonder if you managed to fine-tune the LLM LoRA model on device with onnxruntime?

**Reply by martinkorelic** [reply]
Hi, yes I did manage to develop this as sort of proof of concept (and working on it from time to time nowadays). I was able to deploy and use the fine-tuning with onnxruntime on device by using only its CPU, which ended up [looking something like this](https://martinkorelic.github.io/project/mobiletransformers).

**Reply by xueyedamo521** [reply]
> Hi, yes I did manage to develop this as sort of proof of concept (and working on it from time to time nowadays). I was able to deploy and use the fine-tuning with onnxruntime on device by using only its CPU, which ended up [looking something like this](https://martinkorelic.github.io/project/orttransformer).

Great work! Can you use the fine-tuning with onnxruntime on device by using  mobile GPU?

**Reply by martinkorelic** [reply]
Unfortunately, I could not get this to work with the onnxruntime, I think that currently training on mobile GPU is not supported yet. Tried training with NNAPI, but get an error that training is to use with NNAPI. 
On the other hand, I can’t verify this with QNN since I don’t have access to such hardware. I haven’t checked recently, but if GPU training is supported in ONNX Runtime, it should be straightforward to use with my solution.

**Reply by ColdCodeCool** [reply]
great work!!
